drivers/counter/counter_gd32_timer.c

```diff
@@ -0,0 +1,470 @@
+
+	__ASSERT_NO_MSG(guard < get_top_value(dev));
+
+	data->guard_period = guard;
+	return 0;
+}
+
+static void top_irq_handle(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+	struct counter_gd32_data *data = dev->data;
+	counter_top_callback_t cb = data->top_cb;
+
+	if (timer_interrupt_flag_get(cfg->reg, TIMER_INT_FLAG_UP) == SET) {
+		timer_interrupt_flag_clear(cfg->reg, TIMER_INT_FLAG_UP);
+		__ASSERT(cb != NULL, "top event enabled - expecting callback");
+		cb(dev, data->top_user_data);
+	}
+}
+
+static void alarm_irq_handle(const struct device *dev, uint32_t chan)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+	struct counter_gd32_data *data = dev->data;
+	struct counter_gd32_ch_data *alarm = &data->alarm[chan];
+	const uint32_t reg = cfg->reg;
+	counter_alarm_callback_t cb;
+	bool hw_irq_pending =
+		(timer_interrupt_flag_get(reg, TIMER_FLAG_CH(chan)) == SET);
+	bool sw_irq_pending = data->cc_int_pending & TIMER_INT_CH(chan);
+
+	if (hw_irq_pending || sw_irq_pending) {
+		atomic_and(&data->cc_int_pending, ~TIMER_INT_CH(chan));
+		timer_interrupt_disable(cfg->reg, TIMER_INT_CH(chan));
+		timer_interrupt_flag_clear(cfg->reg, TIMER_FLAG_CH(chan));
+
+		cb = alarm->callback;
+		alarm->callback = NULL;
+
+		if (cb) {
+			cb(dev, chan, timer_counter_read(reg),
+			   alarm->user_data);
+		}
+	}
+}
+
+static void irq_handler(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+
+	top_irq_handle(dev);
+
+	for (uint32_t i = 0; i < cfg->counter_info.channels; i++) {
+		alarm_irq_handle(dev, i);
+	}
+}
+
+static int counter_gd32_timer_init(const struct device *dev)
+{
+	const struct counter_gd32_config *cfg = dev->config;
+	struct counter_gd32_data *data = dev->data;
+	timer_parameter_struct timer_param = {
+		.prescaler = cfg->prescaler,
+		.alignedmode = TIMER_COUNTER_EDGE,
+		.counterdirection = TIMER_COUNTER_UP,
+		.period = cfg->counter_info.max_top_value,
+		.clockdivision = TIMER_CKDIV_DIV1,
+		.repetitioncounter = 0,
+	};
+	uint32_t pclk;
+
+	clock_control_on(GD32_CLOCK_CONTROLLER,
+			 (clock_control_subsys_t *)&cfg->clkid);
+	clock_control_get_rate(GD32_CLOCK_CONTROLLER,
+			       (clock_control_subsys_t *)&cfg->clkid, &pclk);
+	reset_line_toggle_dt(&cfg->reset);
+
+	data->freq = pclk / (cfg->prescaler + 1);
+	cfg->irq_config(dev);
+
+	timer_interrupt_disable(cfg->reg, TIMER_INT_ALL);
+	timer_deinit(cfg->reg);
+
+	gd32_timer_init(cfg->reg, &timer_param);
+
+	return 0;
+}
+
+static const struct counter_driver_api counter_api = {
+	.start = start,
+	.stop = stop,
+	.get_value = get_value,
+	.set_alarm = set_alarm,
+	.cancel_alarm = cancel_alarm,
+	.set_top_value = set_top_value,
+	.get_pending_int = get_pending_int,
+	.get_top_value = get_top_value,
+	.get_guard_period = get_guard_period,
+	.set_guard_period = set_guard_period,
+	.get_freq = get_freq,
+};
+
+#define TIMER_IRQ_CONFIG(n)                                                    \
```

Comment:
Use tab as much as possible to align `\`.

Reply:
The 'clang-format' formatted these lines, so I want it allowed if you could.